derive/src/to_client.rs

```diff
@@ -0,0 +1,134 @@
+                            _ => None,
+                        }
+                    }).collect();
+                method_decl.push(syn::parse_quote! {
+                    fn #name(#args) -> #new_ty;
+                });
+                method_impl.push(syn::parse_quote! {
+                    fn #name(#args) -> #new_ty {
+                        let args_tuple = (#(#arg_names,)*);
+                        let args = serde_json::to_value(args_tuple)
+                            .expect("should never fail");
+                        let params = serde_json::from_value(args)
+                            .expect("should never fail");
+                        let request = Request::Single(Call::MethodCall(MethodCall {
+                            jsonrpc: Some(Version::V2),
+                            method: #rpc_name.to_owned(),
+                            params,
+                            id: Id::Num(next_id()),
+                        }));
+                        let request_str = serde_json::to_string(&request)
+                            .expect("should never fail");
+                        self.call_method(request_str)
+                            .and_then(rpc_impl_response::#name)
+                    }
+                });
+                method_resp.push(syn::parse_quote! {
+                    pub fn #name(response_str: String) -> FutureResult<#returns, Error> {
+                        let response = serde_json::from_str::<Response>(&response_str)
+                            .map_err(|_| Error::new(ErrorCode::ParseError));
+                        let response = match response {
+                            Ok(response) => response,
+                            Err(error) => return future::err(error),
+                        };
+                        let value: Result<Value> = match response {
+                            Response::Single(output) => output.into(),
+                            _ => return future::err(Error::parse_error()),
+                        };
+                        let value = match value {
+                            Ok(value) => value,
+                            Err(error) => return future::err(error),
+                        };
+                        let result = serde_json::from_value::<#returns>(value)
+                            .map_err(|_| Error::new(ErrorCode::ParseError));
+                        match result {
+                            Ok(result) => future::ok(result),
+                            Err(error) => future::err(error),
+                        }
+                    }
+                })
+            }
+            _ => {}
+        }
+    }
+
+    let mut rpc_client_trait = rpc_trait.to_owned();
+    rpc_client_trait.items = method_decl;
+
+    let trait_name = &rpc_client_trait.ident;
+    let rpc_client_impl: syn::ItemImpl = syn::parse_quote! {
+        impl<T: jsonrpc_core::client::RpcClient> #trait_name for T {
```

Comment:
Do we really need a trait for this? While I think it's nice for the server side, when the actual implementation has to be provided by the user, I don't think it's useful in case we generate the implementation anyway.

Maybe let's just stick to regular struct + methods?